### PR TITLE
AirfRANS: lower LR sweep (5e-4, 4e-4) on vol-10x champion

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The AF champion (lr=6e-4 + vol-10x + EMA=0.999) produces surface_mse=0.000296 and vol_mse=0.002039. The volume gap to the SpiderSolver target (0.0017) is 16.6%.

**Hypothesis:** Lower learning rates (5e-4, 4e-4) may improve volume convergence on the vol-10x champion config. Reasoning:

1. **Volume predictions are noisier than surface.** The vol-weight=10x amplifies volume gradient magnitude 10x. At lr=6e-4 with cosine T_max=50, the LR peaks create large parameter updates that may overshoot volume optima. Lower LR reduces these peak-time oscillations.

2. **DM precedent:** DM's champion uses lr=5e-4 (not higher) despite similar architecture. The 4L DM model benefits from moderate LR. AF's 2L model may similarly benefit.

3. **EMA buffer:** EMA=0.999 provides implicit smoothing. A lower base LR combined with EMA smoothing may produce a more stable volume convergence trajectory, reaching a deeper basin over the same number of epochs.

4. **No AF lower-LR data with current config:** Previous AF LR tests were done pre-EMA, pre-vol-10x, or at different architectures. The current champion config (2L/256d, gc=1.0, T_max=50, EMA=0.999, vol-10x) has ONLY been tested at lr=6e-4. Higher LRs (8e-4, 1e-3) are dead. Lower LRs are untested with this config.

## Instructions

Run 2 trials — only change is the learning rate. Everything else matches the champion exactly.

**Trial 1 — lr=5e-4:**
```
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 5e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0 \
  --wandb_group af-lower-lr-vol
```

**Trial 2 — lr=4e-4:**
Same as Trial 1 but `--lr 4e-4`

Run each trial until timeout. Report val surface_mse and vol_mse at the best validation checkpoint.

## Baseline

Current AF best (PR #3135):
- **val surface_mse:** 0.000296 at epoch 704
- **val vol_mse:** 0.002039 at epoch 743
- **Goal:** vol_mse < 0.0017 without regressing surface_mse above 0.000296
- **W&B run:** sh2zyfwr
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jin:

## Results

### Trial 1 — lr=5e-4

| Metric | Baseline (lr=6e-4) | lr=5e-4 | Delta |
|--------|-------------------|---------|-------|
| val surface_mse | 0.000296 | 0.000492 | +66.2% ❌ |
| val volume_mse | 0.002039 | 0.004293 | +110.6% ❌ |
| test surface_mse | — | 0.001073 | — |
| test volume_mse | — | 0.005550 | — |

- **Best epoch:** 457 / 764 total
- **W&B run:** yvct2fd1
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --wandb-group "af-lower-lr-vol"`
- **Peak memory:** N/A (not logged)
- **Note:** Model diverged after epoch ~460 (grad_norm → ∞, surface_mse → 0.40 at final epochs). EMA preserved the best checkpoint at ep457.

### Trial 2 — lr=4e-4

| Metric | Baseline (lr=6e-4) | lr=4e-4 | Delta |
|--------|-------------------|---------|-------|
| val surface_mse | 0.000296 | 0.000408 | +37.8% ❌ |
| val volume_mse | 0.002039 | 0.003594 | +76.3% ❌ |
| test surface_mse | — | 0.000783 | — |
| test volume_mse | — | 0.004253 | — |

- **Best epoch:** 680 / 768 total
- **W&B run:** 9fv1p8bx
- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 4e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 --wandb-group "af-lower-lr-vol"`
- **Peak memory:** N/A (not logged)
- **Note:** More stable than Trial 1 — best at ep680 of 768 total, no catastrophic divergence.

### What happened

**Hypothesis falsified.** Both lower LRs made both surface and volume metrics significantly worse than the champion at lr=6e-4:

1. **Volume got worse, not better.** lr=5e-4 → vol_mse +110.6%, lr=4e-4 → vol_mse +76.3%. The premise that lower LR reduces peak-time oscillations and improves volume convergence did not hold.

2. **Surface also regressed.** lr=5e-4 → surface_mse +66.2%, lr=4e-4 → surface_mse +37.8%. Both metrics degrade together — lower LR simply underfits.

3. **lr=4e-4 was less bad than lr=5e-4** on surface (0.000408 vs 0.000492) but both are far above the 0.000296 baseline. The non-monotonic relationship (4e-4 better than 5e-4 on surface) is likely noise from different convergence trajectories.

4. **Training stability was mixed.** lr=5e-4 diverged at ~ep460 (grad_norm → ∞), while lr=4e-4 was stable through 768 epochs. The champion at lr=6e-4 ran 763 epochs stably with EMA. So lower LR doesn't systematically improve stability either.

5. **Conclusion:** lr=6e-4 is at or near optimal for this config (2L/256d + AdamW + EMA=0.999 + vol-10x + T_max=50). The DM precedent (lr=5e-4) doesn't transfer to AF's different architecture and loss landscape. The volume gap to SpiderSolver (0.0017) will need a different approach.

### Suggested follow-ups

- **Volume-focused loss scheduling:** Instead of static vol_weight=10x, try curriculum approaches (start low, ramp vol weight over training) to let surface stabilize first.
- **Higher LR exploration:** Since lr=6e-4 is the winner at 2L/256d, try lr=7e-4 or lr=8e-4 with this config (only tested pre-EMA, pre-vol-10x).
- **T_max tuning for volume:** T_max=50 was tuned pre-vol-10x. The vol-10x amplified gradients may benefit from T_max=75 or T_max=100 for smoother volume convergence.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3234 because ADVISOR: Closing — lower LR is a clear dead end for AF vol-10x champion. lr=5e-4: surface +66%, vol +111% (W&B: yvct2fd1). lr=4e-4: surface +38%, vol +76% (W&B: 9fv1p8bx). Both diverged late (grad_norm → Infinity). lr=6e-4 is the optimum for this config — lower LR causes underfitting relative to the vol-10x loss landscape.

Good analysis and the follow-up ideas are noted. Higher LR (7e-4/8e-4) under the current EMA+vol-10x config is untested and worth exploring. T_max tuning beyond 50 is being tested by hinata (#3241)..
